packages/storage-plus/src/snapshot_item.rs

```diff
@@ -0,0 +1,290 @@
+pub struct SnapshotItem<'a, T> {
+    primary: Item<'a, T>,
+    snapshots: Snapshot<'a, (), T>,
+}
+
+impl<'a, T> SnapshotItem<'a, T> {
+    /// Usage: SnapshotItem::new(snapshot_names!("foobar"), Strategy::EveryBlock)
+    pub const fn new(
+        storage_key: &'a str,
+        checkpoints: &'a str,
+        changelog: &'a str,
+        strategy: Strategy,
+    ) -> Self {
+        SnapshotItem {
+            primary: Item::new(storage_key),
+            snapshots: Snapshot::new(checkpoints, changelog, strategy),
+        }
+    }
+
+    pub fn add_checkpoint(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
+        self.snapshots.add_checkpoint(store, height)
+    }
+
+    pub fn remove_checkpoint(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
+        self.snapshots.remove_checkpoint(store, height)
+    }
+}
+
+impl<'a, T> SnapshotItem<'a, T>
+where
+    T: Serialize + DeserializeOwned + Clone,
+{
+    /// load old value and store changelog
+    fn write_change(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
+        // if there is already data in the changelog for this block, do not write more
+        if self
+            .snapshots
+            .changelog
+            .may_load(store, ((), U64Key::from(height)))?
+            .is_some()
+        {
+            return Ok(());
+        }
+        // otherwise, store the previous value
+        let old = self.primary.may_load(store)?;
+        self.snapshots
+            .changelog
+            .save(store, ((), U64Key::from(height)), &ChangeSet { old })
+    }
+
+    pub fn save(&self, store: &mut dyn Storage, data: &T, height: u64) -> StdResult<()> {
+        if self.snapshots.should_checkpoint(store, &())? {
+            self.write_change(store, height)?;
+        }
+        self.primary.save(store, data)
+    }
+
+    pub fn remove(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
+        if self.snapshots.should_checkpoint(store, &())? {
+            self.write_change(store, height)?;
+        }
+        self.primary.remove(store);
+        Ok(())
+    }
+
+    /// load will return an error if no data is set, or on parse error
+    pub fn load(&self, store: &dyn Storage) -> StdResult<T> {
+        self.primary.load(store)
+    }
+
+    /// may_load will parse the data stored if present, returns Ok(None) if no data there.
+    /// returns an error on parsing issues
+    pub fn may_load(&self, store: &dyn Storage) -> StdResult<Option<T>> {
+        self.primary.may_load(store)
+    }
+
+    // may_load_at_height reads historical data from given checkpoints.
+    // Only returns `Ok` if we have the data to be able to give the correct answer
+    // (Strategy::EveryBlock or Strategy::Selected and h is registered as checkpoint)
+    //
+    // If there is no checkpoint for that height, then we return StdError::NotFound
+    pub fn may_load_at_height(&self, store: &dyn Storage, height: u64) -> StdResult<Option<T>> {
```

Comment:
This also looks a lot like the `snapshot_map` implementation and most of that could be moved into `Snapshot` (sure, we pass a placeholder (), but saves code duplication of the most important logic.

Reply:
This was non trivial to refactor, because of the coupling with the `primary` map.

Did something that works using nested `Option`. Passing `primary` to `snapshots` is worse I think.